File: 16_Power/Power.py
class Solution(object):
    def Power(self, base, exponent):
        try:
            result = self.power_value(base, abs(exponent))
            if exponent < 0:
                return 1.0 / result
        except ZeroDivisionError:
            print('Error: base is zero')
        else:
            return result

    def power_value(self, base, exponent):
        if exponent == 0:
            return 1
        if exponent == 1:
            return base
        result = self.power_value(base, exponent >> 1)
        result *= result
        if exponent & 0x1 == 1:
            result *= base
        return result

File: 16_Power/Power.cpp
#include <iostream>
#include <vector>
#include <string>
#include <deque>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <algorithm>
#include <ctime>
using namespace std;

bool g_InvalidInput = false;

double PowerWithUnsignedExponent(double base, unsigned int exponent)
{
    if (exponent == 0)
        return 1;
    if (exponent == 1)
        return base;
    double result = PowerWithUnsignedExponent(base, exponent >> 1);
    result *= result;
    if (exponent & 0x1 == 1)
        result *= base;
    return result;
}

double Power(double base, int exponent)
{
    g_InvalidInput = false;
    if (equal(base, 0.0) && exponent < 0)
    {
        g_InvalidInput = true;
        return 0.0;
    }
    unsigned int absExponent = (unsigned int)(exponent);
    if (exponent < 0)
        absExponent = (unsigned int)(-exponent);
    double result = PowerWithUnsignedExponent(base, absExponent);
    if (exponent < 0)
        result = 1.0 / result;
    return result;
}
